tests/codegen/is_val_statically_known.rs

```diff
@@ -0,0 +1,48 @@
+// compile-flags: --crate-type=lib -Zmerge-functions=disabled
+
+#![feature(core_intrinsics)]
+
+use std::intrinsics::is_val_statically_known;
+
+pub struct A(u32);
+pub enum B {
+    Ye(u32),
+}
+
+#[inline]
+pub fn _u32(a: u32) -> i32 {
+    if unsafe { is_val_statically_known(a) } { 1 } else { 0 }
+}
```

Comment:
Isn't tests here UB?
https://llvm.org/docs/LangRef.html#id1676:
```
The result also intentionally depends on the result of optimization passes – e.g., the result can change depending on whether a function gets inlined or not.
```

Results here can depend on opt level/llvm version/etc; no one forbids llvm optimizer became dumb one day, but this shouldn't break code?

Reply:
How is that [UB](https://doc.rust-lang.org/reference/behavior-considered-undefined.html) though? It just means it's non-deterministic. But here we are testing that it works as expected for this particular configuration.

Reply:
That is only incorrect if it is in a `const` context or `const` function. Even in a `const` function I'm not sure it is really UB, just unsound.

Reply:
@RalfJung Could this function be safe if given a wrapper that checks types and isn't `const`?

Reply:
Yeah this function could probably be safe. It's just a non-deterministic `bool`.

Seems better to first land the MVP though. I think we'll want to allow this safely even in const but that's a [different discussion](https://github.com/rust-lang/rfcs/pull/3352).

Reply:
Great! I meant that as a hypothetical though.